sdk/monitor/Azure.Monitor.OpenTelemetry/src/AzureMonitorOpenTelemetryImplementations.cs

```diff
@@ -0,0 +1,177 @@
+                             .AddAzureMonitorTraceExporter(o =>
+                             {
+                                 o.ConnectionString = options.ConnectionString;
+                                 o.DisableOfflineStorage = options.DisableOfflineStorage;
+                                 o.StorageDirectory = options.StorageDirectory;
+                             }));
+            }
+
+            if (options.EnableMetrics)
+            {
+                builder.WithMetrics(b => b
+                             .AddAspNetCoreInstrumentation()
+                             .AddAzureMonitorMetricExporter(o =>
+                             {
+                                 o.ConnectionString = options.ConnectionString;
+                                 o.DisableOfflineStorage = options.DisableOfflineStorage;
+                                 o.StorageDirectory = options.StorageDirectory;
+                             }));
+            }
+
+            return services;
+        }
+
+        internal static IServiceCollection AddAzureMonitorOpenTelemetryWithAction(IServiceCollection services, Action<AzureMonitorOpenTelemetryOptions> configureAzureMonitorOpenTelemetry, string? name)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            name ??= Options.DefaultName;
+
+            if (configureAzureMonitorOpenTelemetry != null)
+            {
+                services.Configure(name, configureAzureMonitorOpenTelemetry);
+            }
+
+            var builder = services.AddOpenTelemetry();
+
+            builder.WithTracing(b => b
+                            .AddAspNetCoreInstrumentation()
+                            .AddAzureMonitorTraceExporter());
+
+            builder.WithMetrics(b => b
+                            .AddAspNetCoreInstrumentation()
+                            .AddAzureMonitorMetricExporter());
+
+            ServiceDescriptor? sdkTracerProviderServiceRegistration = null;
+            ServiceDescriptor? sdkMeterProviderServiceRegistration = null;
+
+            foreach (var service in services)
+            {
+                if (service.ServiceType == typeof(TracerProvider))
+                {
+                    sdkTracerProviderServiceRegistration = service;
+                }
+                else if (service.ServiceType == typeof(MeterProvider))
+                {
+                    sdkMeterProviderServiceRegistration = service;
+                }
+            }
+
+            if (sdkTracerProviderServiceRegistration?.ImplementationFactory == null ||
+                sdkMeterProviderServiceRegistration?.ImplementationFactory == null)
+            {
+                throw new InvalidOperationException("OpenTelemetry SDK has changed its registration mechanism.");
+            }
+
+            // We looped through the registered services so that we can take over
```

Comment:
is this done to remove any previously added providers and only do the one via distro?

Reply:
It is done to read the values from `options.EnableMetrics` or `options.EnableTraces` and act according to it. We get the options as Action<AzureMonitorOpenTelemetryOptions> configureAzureMonitorOpenTelemetry and value for it will not be available until Service Collection is built.